[FAB-17059]: Change collection membership eligibility checks to only compare mspID

caod123 opened the pull request:
#### Type of change

- Bug fix

#### Description

This attempts to fix a bug where eligibility evaluation was failing when ca certs were rolled

#### Additional details

See [FAB-17059](https://jira.hyperledger.org/browse/FAB-17059)

/cc @ale-linux

caod123 committed:
[FAB-17059] Change collection membership eligibility checks to
only compare mspID instead of Evaluating the certificate against the
access policy

* This handles the issue when ca certs are rolled and access policy
evalutaion for eligible private data was failing due to policy
evaluation checking certs prior to the config update block that added
the new cert

Signed-off-by: Danny Cao <dcao@us.ibm.com>

yacovm commented on gossip/service/gossip_service.go:
```diff
@@ -111,6 +111,7 @@ type gossipServiceImpl struct {
 	lock            sync.RWMutex
 	mcs             api.MessageCryptoService
 	peerIdentity    []byte
+	mspID           string
```
you can extract the MSPID by taking the `peerIdentity` and passing into the `secAdv` that is right below this `mspID` field and casting it to string.

caod123 replied:
Thanks, I was trying to find a way of extracting the mspID without having to pass it down the entire stack. Any other thoughts about this approach? I tried to ask @ale-linux but haven't heard back yet.

caod123 replied:
Also that covers extracting the mspID for the gossip path -> coordinator. Is there an easier way of extracting from the membership update path via the `membershipInfoProvider`? I'd like to entirely do away with the `selfSignedData` in both paths since I think we're only using them for the existing policy evaluation logic, but I guess for now we can keep them for the sake of falling back in case the new method doesn't work.

yacovm replied:
You can apply the same function of the `secAdv` to the Identity in the self signed data, of course

caod123 replied:
That works for the gossip path since it has a `secAdv` but the `membershipInfoProvider` only takes an `IdentityDeserializerFactory` and `SignedData`

yacovm commented on peer/node/start.go:
```diff
@@ -871,7 +873,7 @@ func secureDialOpts() []grpc.DialOption {
 // 3. Init the security advisor;
 // 4. Init gossip related struct.
-func initGossipService(policyMgr policies.ChannelPolicyManagerGetter, metricsProvider metrics.Provider,
+func initGossipService(mspID string, policyMgr policies.ChannelPolicyManagerGetter, metricsProvider metrics.Provider,
```
why do you need to put here this mspID in this `initGossipService` ? 

You have the identity passed (`serializedIdentity`) and below you also have the `secAdv` as well.

caod123 replied:
ah oops I missed this one, I meant to remove all of the passed mspIDs in the gossip path after we discussed the `secAdv` but forgot about this one. Thanks for catching it!

ale-linux commented on core/common/privdata/membershipinfo.go:
```diff
@@ -7,36 +7,98 @@ SPDX-License-Identifier: Apache-2.0
+		memberOrgs, err := getMemberOrgs(collectionPolicyConfig.GetSignaturePolicy().GetIdentities(), deserializer)
+		if err != nil {
+			logger.Errorf("Failed getting member orgs from MSP Principals: %s", err)
+			return false, nil
```
Why don't we fall back here?

ale-linux commented on core/common/privdata/membershipinfo.go:
```diff
@@ -7,36 +7,98 @@ SPDX-License-Identifier: Apache-2.0
+				return true, nil
+			}
+		}
+		return false, nil
```
and here? In these 2 cases I would just let `getPolicy` make the decision for us

ale-linux reviewed (changes requested):
Thanks Danny! One thing that I'm still missing is why the two separate paths (one through AmMemberOf and the other through isEligible. I would also add some additional comments in the code explaining why we're doing what we're doing. Finally, we'll need some kind of integration tests for this particular scenario.

C0rWin commented on core/common/privdata/membershipinfo.go:
```diff
@@ -7,36 +7,98 @@ SPDX-License-Identifier: Apache-2.0
+
+// Returns a list of member orgs from a list of MSPPrincipals
+func getMemberOrgs(identities []*mspp.MSPPrincipal, deserializer msp.IdentityDeserializer) ([]string, error) {
+	memberOrgs := []string{}
```
a nit, but doesn't it make sense to have here a map from string to bool/struct{}? this should make your later check for inclusion easier.

caod123 replied:
Good call thanks, I'll make the change

denyeart commented on core/common/privdata/membershipinfo.go:
```diff
@@ -7,36 +7,98 @@ SPDX-License-Identifier: Apache-2.0
 // AmMemberOf checks whether the current peer is a member of the given collection config.
 // If getPolicy returns an error, it will drop the error and return false - same as a RejectAll policy.
-func (m *MembershipProvider) AmMemberOf(channelName string, collectionPolicyConfig *common.CollectionPolicyConfig) (bool, error) {
+func (m *MembershipProvider) AmMemberOf(channelName string,
```
AmMemberOf is used when a chaincode is upgraded, to see if the peer's org has become eligible after a collection change. May want to indicate this in the comment.

denyeart commented on core/common/privdata/membershipinfo.go:
```diff
@@ -7,36 +7,98 @@ SPDX-License-Identifier: Apache-2.0
 	return true, nil
 }
+
+// Returns a list of member orgs from a list of MSPPrincipals
```
Comment should start with "getMemberOrgs" right?

denyeart commented on gossip/privdata/coordinator.go:
```diff
@@ -824,6 +826,14 @@ func (c *coordinator) accessPolicyForCollection(chdr *common.ChannelHeader, name
 
 // isEligible checks if this peer is eligible for a given CollectionAccessPolicy
 func (c *coordinator) isEligible(ap privdata.CollectionAccessPolicy, namespace string, col string) bool {
+	// Simple check to see if mspID part of collection's MemberOrgs list
```
isEligible is used upon commit to determine if the peer is eligible to retrieve/persist the private data collection data for a given transaction. May want to indicate this in the comment.

denyeart reviewed (commented):
Agree with the other feedback, let's add as many comments as we can so that the behavior becomes clear.

denyeart commented:
I've tested in an e2e scenario with some additional debug, and confirmed the new simple checks are working as expected on the private data requester side, but not called on the private data provider side (provider will continue to use the policy evaluation).

caod123 committed:
[FAB-17059] Extend private data integration tests to cover case when a
new eligible peer with different certs joins the channel and attempts to fetch
private data before processing the config update that adds its certs to
the channel config

Signed-off-by: Danny Cao <dcao@us.ibm.com>

caod123 commented:
The integration test ended up being fairly complicated. Please let me know if you have questions or concerns regarding the reproduction case. If desired I can squash into 1 commit. Once this PR is merged I'll port to 2.1

cendhu commented on core/common/privdata/membershipinfo.go:
```diff
@@ -7,36 +7,97 @@ SPDX-License-Identifier: Apache-2.0
+			mspRole := &mspp.MSPRole{}
+			err := proto.Unmarshal(principal.Principal, mspRole)
+			if err != nil {
+				return memberOrgs, errors.Wrap(err, "Could not unmarshal MSPRole from principal")
```
I think the error message should start with a small letter though it is not explicitly mentioned in the [coding guidelines](https://hyperledger-fabric.readthedocs.io/en/release-1.4/error-handling.html).

cendhu commented on core/common/privdata/membershipinfo.go:
```diff
@@ -7,36 +7,97 @@ SPDX-License-Identifier: Apache-2.0
+	memberOrgs := make(map[string]struct{})
+
+	// get member org MSP IDs from the envelope
+	for _, principal := range identities {
```
Did we take this code from [core/common/privdata/simplecollection.go](https://github.com/hyperledger/fabric/blob/master/core/common/privdata/simplecollection.go#L112)? If so, why not move this common code to [util.go](https://github.com/hyperledger/fabric/blob/master/core/common/privdata/util.go) and avoid duplication within a package?

caod123 replied:
Yep, sure I can move it to a shared place. I left it separate because we ended up changing it to return a map instead of a string slice. I can change the consumers of SimpleCollection (coordinator) to also check the map instead.

cendhu commented on core/common/privdata/membershipinfo.go:
```diff
@@ -7,36 +7,97 @@ SPDX-License-Identifier: Apache-2.0
 // AmMemberOf checks whether the current peer is a member of the given collection config.
 // If getPolicy returns an error, it will drop the error and return false - same as a RejectAll policy.
-func (m *MembershipProvider) AmMemberOf(channelName string, collectionPolicyConfig *common.CollectionPolicyConfig) (bool, error) {
+// It is used when a chaincode is upgraded to see if the peer's org has become eligible after	a collection
```
A nit. Not sure whether we need to comment about the usages of a function. We have tools/editors that help in finding all the callers/callee/references. I am saying this because we use this function twice during a chaincode upgrade.
1.  to check whether the peer is already a member (using the existing collection policy);
2.  to check whether the peer is a newly added member (using the new policy if the check during (1) returns false).

Using both (1) and (2), we check whether a peer has become a new member of a collection. I understand that this comment was added as a result of some questions raised during the review (when `AmMemberOf` and `isEligible` are used). Hence, I am fine if we plan to retain it.

denyeart replied:
I've confused between AmMemberOf and isEligible, so I'm fine to keep it.

cendhu commented on core/common/privdata/collection.go:
```diff
@@ -29,7 +29,7 @@ type Collection interface {
 
 	// MemberOrgs returns the collection's members as MSP IDs. This serves as
 	// a human-readable way of quickly identifying who is part of a collection.
-	MemberOrgs() []string
```
We can still keep it as `[]string` as the goal is to minimize changes made to release 1.4. We have two other options: 

1. we can convert the map to `[]string` at https://github.com/hyperledger/fabric/blob/1007a5f21fb6b2821535a4e23f2acfe9924122af/core/common/privdata/simplecollection.go#L89
2. we can return `[]string` from `getMemberOrgs` instead of a map. We can do a linear search on the `[]string` to find the match. Given that we use this function only during the chaincode upgrade, no optimization (usage of the map) is required IMO. If optimization is required (though it is not), we in fact need to check for the match after each deserialization of identity rather than collecting all identities at once and then do the matching. 

I think we can go with option (2).

caod123 replied:
I believe that was what it was before, I originally changed to a map via an earlier suggestion. I do see that the extent of the change is a little too broad for 1.4 though so I'll probably just do the map in the 2.0/master branch fix.

cendhu reviewed (commented):
Thanks, @caod123 , for the changes.

yacovm commented on integration/pvtdata/pvtdata_test.go:
```diff
@@ -230,38 +358,90 @@ var _ bool = Describe("PrivateData", func() {
+						Expect(err).NotTo(HaveOccurred())
+						Eventually(sess, network.EventuallyTimeout).Should(gexec.Exit(0))
+					} else {
+						sess, err = network.PeerUserSession(peer, "User1", commands.ChannelInfo{
```
if you have "getLedgerHeight" why do you need this "ChannelInfo" command? can't you use the getLedgerheight ?

caod123 replied:
getLedgerHeight specifically uses `User1` for the peer user session. Unfortunately any operations on the new peer must use the `Admin2` user so a lot of the logic in this integration test revolves around passing in a custom user for the peer user session commands that need to be run on `org2.peer1`. Because this seemed like a pretty isolated test case I figured it wasn't worth extending the functions to take a user in and just leave them to default to `User1`

yacovm replied:
1. Why can't you use User1 ? 
2. Can't you add a parameter to the **getLedgerHeight** function to distinguish?

caod123 replied:
User1 certs can't be replaced because org2.peer0 still uses them, instead I copied the generated admin user over as Admin2 specifically to use for org2.peer1 operations. I don't mind extending `getLedgerHeight` but figured it wasn't worth it for just this use case.

yacovm commented on integration/pvtdata/pvtdata_test.go:
```diff
@@ -199,29 +247,109 @@ var _ bool = Describe("PrivateData", func() {
+			oldMspConfig.RootCerts = append(oldMspConfig.RootCerts, newMspConfig.RootCerts...)
+			oldMspConfig.TlsRootCerts = append(oldMspConfig.TlsRootCerts, newMspConfig.TlsRootCerts...)
+			oldMspConfig.FabricNodeOus.PeerOuIdentifier.Certificate = nil
+			oldMspConfig.FabricNodeOus.ClientOuIdentifier.Certificate = nil
```
all this huge block where you edit the MSPConfig can and should be extracted to its own function.

cendhu commented:
Thanks, Danny, for the changes.

ale-linux commented on core/common/privdata/membershipinfo.go:
```diff
@@ -16,27 +16,50 @@ var logger = flogging.MustGetLogger("common.privdata")
+			logger.Errorf("Failed getting member orgs from MSP Principals: %s", err)
+		}
+
+		for _, org := range memberOrgs {
```
Just to double check that I got it right: getMemberOrgs may still return a non-empty slice of orgs even in case of error, and we're still willing to look at them, in spite of the error. If that's the case, shouldn't getMemberOrgs skip over - say - an error at principal i, to try and see if any principal following can still be appropriately interpreted?

caod123 replied:
Yes, you got that right. I agree with your interpretation of what we should do with getMemberOrgs, I'll have it just continue the loop instead of returning on error. In that case I'll probably also just remove the error from being returned from `getMemberOrgs`.

ale-linux commented on core/common/privdata/util.go:
```diff
@@ -35,3 +39,38 @@ func getPolicy(collectionPolicyConfig *common.CollectionPolicyConfig, deserializ
+			return memberOrgs, errors.New(fmt.Sprintf("invalid principal type %d", int32(principal.PrincipalClassification)))
+		}
+	}
+	return memberOrgs, nil
```
See comment above, if we return a non-empty slice in case of error, I think we should just skip over principals from which we can't extract an MSP ID to see if we can get something out of each principal in the slice.

ale-linux commented on integration/pvtdata/pvtdata_test.go:
```diff
@@ -199,29 +247,105 @@ var _ bool = Describe("PrivateData", func() {
 			org2peer1 := network.Peer("org2", "peer1")
-			network.JoinChannel("testchannel", orderer, org2peer1)
-			org2peer1.Channels = append(org2peer1.Channels, &nwo.PeerChannel{Name: "testchannel", Anchor: false})
+			oldPeerMSPPath := filepath.Join(network.RootDir, "crypto", "peerOrganizations", "org2.example.com", "peers", "peer1.org2.example.com", "msp")
```
Picky: Could you please break up these lines and the other very long ones below?

caod123 replied:
This should have just been `PeerLocalMSPDir(org2peer1)` in the first place. I'll break up the one below it

ale-linux commented on integration/pvtdata/pvtdata_test.go:
```diff
@@ -199,29 +247,105 @@ var _ bool = Describe("PrivateData", func() {
+				filepath.Join(oldAdminUserPath, "msp", "signcerts", "Admin@org2.example.com-cert.pem"),
+				filepath.Join(oldAdminUserPath, "msp", "signcerts", "Admin2@org2.example.com-cert.pem"),
+			)
+			Expect(err).NotTo(HaveOccurred())
```
Can we bubble lines 256-285 out in a helper function?

ale-linux commented on integration/pvtdata/pvtdata_test.go:
```diff
@@ -199,29 +247,105 @@ var _ bool = Describe("PrivateData", func() {
+				Type:   oldConfig.Type,
+				Config: utils.MarshalOrPanic(mergedMspConfig),
+			})
+			nwo.UpdateConfig(network, orderer, "testchannel", currentConfig, updatedConfig, false, network.Peer("org2", "peer0"))
```
Same as the comment above, I wouldn't inline the plumbing but just create helper functions and call them to perform the significant steps, so that someone reading the test would understand the test case

ale-linux reviewed (dismissed):
Thanks Danny! Some nits from my side
